### ISSUE TITLE
Joint consensus not implemented correctly

### ISSUE THREAD

**Issue by Diggsey** [railgun-rs/actix-raft#62]
I believe votes are counted incorrectly during joint consensus.

Firstly, during this phase, votes should be counted separately for the two clusters. Quote from the spec:
> Agreement (for elections and entry commitment) requires separate majorities from both the old and new configurations.

Currently the code only tracks a single vote counter in the candidate state.

Secondly, this is not so clearly explained, but I believe a candidate which will be removed after a membership change, which has received the log entry containing the membership change, should only count the vote for itself in the "old" vote counter, but not the "new" vote counter.

**Comment by Diggsey** [comment]
If it helps, I've been modelling membership like this:
```rust

/// Each voting group reaches consensus independently. Nodes may belong to multiple
/// voting groups.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct VotingGroup {
    pub members: HashSet<NodeId>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Membership {
    /// During joint consensus, there will be two voting groups. The old group will be first, followed
    /// by the new group.
    pub voting_groups: Vec<VotingGroup>,
    /// List of nodes which are not part of a voting group.
    pub non_voters: HashSet<NodeId>,
}
```

I then keep one vote counter per voting group:
```rust

// Counts votes within a single voting group
#[derive(Default, Clone)]
struct MajorityCounter {
    votes_received: u64,
    has_majority: bool,
}

impl MajorityCounter {
    fn add_vote(&mut self, from: NodeId, group: &VotingGroup) {
        if group.members.contains(&from) {
            self.votes_received += 1;
            if self.votes_received > (group.members.len() as u64) / 2 {
                self.has_majority = true;
            }
        }
    }
}

struct CandidateState {
    // One vote counter per voting group
    vote_counters: Vec<MajorityCounter>,
}
```

This greatly simplifies edge cases like not counting votes for yourself if you're not part of one of the voting groups, since even self-votes go through the `add_vote` method.

**Comment by thedodd** [comment]
So, the implementation of joint consensus in this system is optimized and simplified a bit. Membership configuration is always expressed as a unified configuration, with the actual change to the config being expressed in terms of nodes being added and nodes being removed.

All of the actual safety requirements mandated by the spec are still upheld. At the end of the day, this is just a different way of modelling the config. The spec doesn't actually described a data model for the config.

Also, just think of this purely in terms of mathematical sets. Both `C0` & `C1` (in the language of the spec) are expressed in this implementation's "unified config" model.
- there can only be one leader elected, and it must be the same from `C0` & `C1`.
- instead of tracking a `C0` & a `C1` for joint consensus, we track a merged and unified config which describes the action being taken on the cluster's config.
- when joint consensus is started, a boolean value in the config is set to `true` which indicates that we are in the first part of the joint consensus two-phases commit process.
- after the initial joint consensus is committed, the final part of the two-phase commit will apply the original actions to the config, set the joint consensus bool back to `false` and will then commit that config.
- when a config change is proposed to the cluster, the leader node validates the proposed config change before just blindly accepting it. Safety guaranties are enforced on the config. For example:
    - proposing a config change which removes all nodes from the cluster ... obviously not going to fly.
    - proposing to remove all nodes an leave only one, also not going to fly because that would compromise the core data safety guaranties of Raft.
    - these validations overall ensure that the unified configuration set will always produce a safe cluster membership from which a new leader can always be elected.

So, the spec requires `separate majorities from both the old and new configurations`. That is accomplished by this system, it is just modeled differently. The fact that we model the action taken, but still express the set of all nodes, accomplishes this goal.

There are subtle cases which one might imagine where crazy configuration updates are taking place. Those cases are fine, as long as one remembers the fundamentals required by raft during these phases:
- new nodes being added are "non-voters" until they are brought up-to-speed and then the final config of the two-phase commit is committed. Until then, the new nodes are still just non-voters. They don't vote. They do not campaign for leadership.
    - a campaign taking place during the joint phase (C0C1 phase) of a config change can only select from the nodes of C0, by definition. New nodes of C1 literally will not even start a campaign.
- once C1 is committed (the final phase of the two-phase commit), the entire set of nodes expressed in C0C1 are still replicated to.
    - added nodes which were non-voters in C0C1, are not considered voting members in C1.
    - nodes which have been removed and which no longer exist in C1 are still replicated to until C1 is committed to the cluster and until the target nodes themselves have replicated C1, at which time those nodes will go into non-voter state and the leader will close their replication streams.

**In conclusion,** the joint consensus system in this implementation is implemented correctly, it just models the configuration in a way which is different that what the spec describes; though, as mentioned earlier, the spec does not mandate an exact structure/model for the membership config. So this is, in my opinion, an acceptable model.

The guiding light here is: **be sure that we uphold Raft's safety guaranties!** This will always be implementation specific.

**Comment by Diggsey** [comment]
I'm open to being convinced otherwise, but I still don't believe that it's possible to accurately determine separate majorities without separate vote counters.

The first issues I see is with this statement:
> there can only be one leader elected, and it must be the same from C0 & C1.

The transition from C0 -> C0C1 is not atomic. It takes time for the new configuration to be replicated. During this time, nodes from C0 which have not received the config change can still elect a node which belongs to C0 but not C1.

With a single counter, a vote from a given node can either be counted or not counted:

- If you count votes from both C0 and C1 nodes, then you could end up replicating to a majority of nodes in C0 but not in C1, or vice versa. If the latter case happens (you replicate an entry to a majority of C1 nodes, but not C0) then a C0 node without that entry can later be elected and overwite the (supposedly committed) entry.

- If you count votes only from nodes in both C0 and C1, then the cluster becomes significantly less fault-tolerant: it may only barely be possible to get an overall majority from that set, and so a single failure can cause the system to become unavailable. For example, if C0 = {A, B, C} and C1 = {A, B, D} then during joint consensus a failure of either A or B will cause the cluster to become stuck. In contrast, with separate vote counters, both configurations can still reach consensus with any single node failure.

There are symmetrical issues on the other side, when leaving joint consensus.

The second issue I see is with this statement:
> new nodes being added are "non-voters" until they are brought up-to-speed and then the final config of the two-phase commit is committed. Until then, the new nodes are still just non-voters. They don't vote. They do not campaign for leadership. 

This is at odds with multiple statements from the spec: non-voters should become voters as soon as the C0C1 configuration change is replicated to them (before it is even committed):

> Once a given server adds the new configuration entry to its log, it uses that configuration for all future decisions (a server always uses the latest configuration in its log, regardless of whether the entry is committed).

The non-voter phase is an entirely separate phase that happens before the C0C1 configuration is even replicated:

> In order to avoid availability gaps, Raft introduces an additional phase **before the configuration change**, in which the new servers join the clusters non-voting members

In other words:

1) New servers are added as non-voters to the cluster. It is irrelevant whether this entry is committed or not. It doesn't even need to *be* a log entry - non-voters could be managed separately as they can be added and removed without concern, and it doesn't even matter whether all nodes have a consistent set of non-voters. All that really matters is that these new servers are caught up *somehow* to the state of the cluster.

2) The leader proposes a membership change by starting replication of C0C1

3) C0C1 is committed (replicated to a majority of both C0 and C1)

4) The leader starts replication of C1

5) C1 is committed (replicated to a majority of C1)

During phases 2, 3 and 4, the system is in joint consensus, neither cluster can operate independently. Up until phase 5, a leader from C0 can still be elected.

Finally, there's also the issue of not counting votes for yourself if you're not part of the configuration, which is called out in the spec:
> This means that there will be a period of time (while it is committing Cnew) when the leader is managing a cluster that does not include itself; it replicates log entries but does not count itself in majorities.

This is not handled in the code from what I could tell.

**Comment by thedodd** [comment]
Well ... I was in the process of writing out my rebuttal, and showed that there was no possibility of a safety violation if an election were to take place during `C0C1` (due to some aspects of how I've implemented this so far). However, in that process, I was able to prove a safety violation which could take place while committing `C1`. So yea, @Diggsey I'm glad that you've brought this up and have challenged my thinking on the current implementation.

I'll review your suggested `VoteGroup` pattern above, and I do think it would be better for me to update the config pattern to stick more closely to the data model which seems to be suggested in the spec.

Thanks for the discussion so far, I'll circle back shortly.

**Comment by Diggsey** [comment]
I spent some time looking into the worst case scenarios during phases 2 and 4. I found it easiest to think about in terms of "added", "shared" and "removed" nodes.

**Phase 2 worst case scenario:** C0C1 is replicated to all removed and shared nodes, but leader crashes before it can be replicated to any added nodes. To achieve consensus within C1, the shared nodes must outnumber the (unaware) added nodes:

**Safety condition 1:** num_shared - fault_tolerance > num_added

**Phase 4 worse case scenario:** C1 is replicated to all removed nodes, but leader crashes before it can be replicated to any shared or added nodes. To achieve consensus within C0, the shared nodes must outnumber the (unaware) removed nodes:

**Safety condition 2:** num_shared - fault_tolerance > num_removed


One interesting consequence of this is that safely replacing a node in a 3-node cluster must be done with multiple configuration change cycles:
1) Add the new node (num_shared=3, num_removed=0, num_added=1)
2) Remove the old node (num_shared=3, num_removed=1, num_added=0)

And this means there is actually value in having a even-sized clusters, as even-sized clusters allow for more efficient membership changes without reducing fault tolerance.

Furthermore, the changes necessary to safely get from any configuration A to configuration B can be determined fairly easily using a greedy algorithm: take the desired set of additions and removals, and cancel additions and removals until both inequalities are satisfied. Cancelling either an "addition" or a "removal" brings condition 1, one step closer to being satisfied. Cancelling a "removal" brings condition 2, two steps closer to being satisfied.

**Comment by Diggsey** [comment]
It seems this problem ^ can be avoided as long as non-voters continue to vote (but the candidate must check that votes it receives are from the relevant voting group before counting that vote). The tikv implementation calls them "learners" instead of "non-voters" as the latter name is obviously incredibly misleading.

With this change to the algorithm, the only inequality that must be satisfied is:

- num_shared > fault_tolerance

Which means you can get from any set of nodes A to any other set of nodes B wtih at most 3 membership change cycles.

**Comment by thedodd** [comment]
BTW, I've completely done away with the way that membership is modeled in 4.x `:)`. Over in #59 I've been updating the membership system as follows:
- the API for config changes now has three methods: `add_member`, `remove_member` & `reconfigure`. The first two allow for applications with dynamic peer discovery to be able to operate in a more fluid manner. Requesting to add nodes as they become available, and to remove nodes for self-healing and such. The `reconfigure` method is for more manual config updates. Perhaps for applications which use an admin CLI for controls &c.
- new nodes are just tracked by the leader, and are synced. They are not used for votes for commit counts. They are replicated to only for the purpose of bringing them up-to-speed. Once they are ready, the leader will commit a new config entry which includes the new node as a voter, and then everything is good from there.
- per our discussion above, commit counts & votes during joint consensus is being updated per the spec's guidelines (separate majorities and such).
